CMake/build.cmake

```diff
@@ -13,23 +13,27 @@ if(NOT CMAKE_BUILD_TYPE)
   set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
 endif()
 
+# SENSEI should always be position independent
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+
+# SENSEI requires minimum C++11
+if (NOT CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD LESS 11)
+  set(CMAKE_CXX_STANDARD 11)
+endif ()
+
 if (NOT MSVC)
-  if (NOT CMAKE_CXX_FLAGS)
-  set(tmp "-fPIC -std=c++11 -Wall -Wextra")
+  string(APPEND CMAKE_CXX_FLAGS "-Wall -Wextra")
+
   if (BUILD_STATIC_EXECS)
-    set(tmp "${tmp} -static -static-libgcc -static-libstdc++ -pthread -Wl,-Bstatic")
-  endif()
-  if ((APPLE) AND ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang"))
-    set(tmp "${tmp} -stdlib=libc++")
-  endif()
-  if ("${CMAKE_BUILD_TYPE}" MATCHES "Release")
-    set(tmp "${tmp} -O3 -march=native -mtune=native")
```

Comment:
Was there a specific problem with these flags?

I would like to keep  `-O3 -march=native -mtune=native` since provides reasonable default for optimized builds. SENSEI prioritizes performance over portability, and  -O3 enables a number of important optimizations such as loop vectorization, which is much more effective when using processor specific capability. This is an important one for GCC, but some other compilers such as Intel have vectorization enabled at -O2.

Reply:
CMake specifies `-O3` by default for builds using Release. It also turns off debug and is cross-platform/compiler compatible which hard-coding is not.

I removed the the other two options because they are not supported by all compilers and lead to build failures. They also do not change the default settings for the compiler when they are valid so effectively they are doing nothing but making the build not portable.

I am open to other solutions if you have them, but this was the fastest way forward I saw that would have the little to no impact on existing downstream projects.

Reply:
/global/common/software/m636/builds/SENSEI/build/lib/cmake/  The problem we were seeing in WarpX (and AMReX) was that the way these flags were getting set was polluting upstream builds. The breaking one in AMReX and WarpX is the `-std=c++11` flag which prevented AMReX and WarpX from building without specifically setting C++14 during configuration.

Reply:
@kwryankrattiger Which compiler does this not work with?  Would conditionally excluding for that compiler be an option?

Reply:
> CMake specifies -O3 by default for builds using Release. It also turns off debug and is cross-platform/compiler compatible which hard-coding is not.

@kwryankrattiger I'm observing that the defaults are actually `-O2`(see output from below), and hence the need to specify `-O3`. Also, note that `CMAKE_CXX_FLAGS` are unioned with the other flags. When compiling in release mode,  `CMAKE_CXX_FLAGS_RELEASE` are also used and `-DNDEBUG` is obtained from there.

Also, I think we should keep the pattern of building the flags in a temporary and setting the `CMAKE_CXX_FLAGS` flags with CACHE FORCE. One good reason for this is that with clang `-stdlib=libc++` is required and I believe due to how CACHE variables work in CMake that there are some cases where if we don't FORCE it, the variable will not be updated.


some output that verify CMake's defaults
```
$cat ../CMakeLists.txt 
project("test_cmake")
message(STATUS "Test!")
```

```
$rm -rf *; cmake .. -LA | grep CMAKE_CXX_FLAGS
CMAKE_CXX_FLAGS:STRING=
CMAKE_CXX_FLAGS_DEBUG:STRING=-g
CMAKE_CXX_FLAGS_MINSIZEREL:STRING=-Os -DNDEBUG
CMAKE_CXX_FLAGS_RELEASE:STRING=-O2 -DNDEBUG
CMAKE_CXX_FLAGS_RELWITHDEBINFO:STRING=-O2 -g -DNDEBUG
```

Reply:
> Interesting, I see it producing -O3

It's not using -O3 on my system as the above output shows. I did the above test on 3 other systems that I can log into and it also shows -O2 on 1 of them and -O3 on 2. The 2 systems that show -O2 are Fedora based systems. Perhaps it has to do with how package managers are building CMake (or gcc?) on Fedora? However, it doesn't matter. the fact that it is inconsistent across various systems, means we need to set it our selves. It is important to performance to have it enabled.

CACHED values only get modified if they are not already set. This is true whether or not using set or append, correct?

Unfortunately, with a required argument the only way to make sure it gets into the flags is to force it, since there are cases when appending it will not modify the value.

GNU, Intel, and Clang all support -march=native -mtune=-native. I would not be surprised if some other compiler doesn't support these, but I don't know that we need to care about that compiler. I like your idea of testing for support. Either way I think it should use the pattern that was there before of filling a temporary and set+force. I'm not keen to change this because there is not a specific issue you're addressing and the changes would create a new issue (compiler vectorization being disabled) on some of the systems we run on.

>One issue I have run into with sensei cmake is CACHE FORCE is being used to cache things that needn't be, to fix those issues I end up having to delete the entire build tree between updates which is why I removed the CACHE FORCE here.

Since we need to modify a CAHCE value, and FORCE is the only way to be sure that modification always happens, we need to FORCE this set.

For better or worse, needing to `rm -rf *` in the build dir to wipe out cached values is a well known pattern in the cmake workflow. It is inherent in CMake design and not particular to sensei.

Reply:
Using a cache variable _requires_ the `FORCE` option. By setting it locally with `string(APPEND)` we circumvent the cache. CMake will prefer locally set variables over CACHE variables, by using `string(APPEND)` we will append to the variable initialized by the cache, allowing for options to be updated between configurations without having to delete everything and re-configure.

Here is an example I generated by setting the cache to `"hello"` and then running configure again

```
-- CMAKE_CXX_FLAGS="hello"-Wall -Wextra -mtune=native -march=native
```
A better way to think about CACHE is the _initial_ option for a variable, and then the local variables in CMake are what are used in the build. This is a subtle difference in CMake that I too just learned about recently.

As far as the options go, I will explicitly edit the `CMAKE_CXX_FLAGS_RELEASE` with  `-O3 -DNDEBUG -mtune=native -march=native`. I couldn't find anything that was convincing as to why CMake would lower the optimization level based on an arch/OS.